src/drivers/dw/dma.c

```diff
@@ -352,11 +352,11 @@ static int dw_dma_release(struct dma_chan_data *channel)
 	/* get next lli for proper release */
 	dw_chan->lli_current = (struct dw_lli *)dw_chan->lli_current->llp;
 
+	irq_local_enable(flags);
+
 	/* prepare to start */
 	dw_dma_stop(channel);
 
-	irq_local_enable(flags);
```

Comment:
no, this breaks atomicity. The original code made sure that no IRQ could interrupt execution between line 353 and the call to `dw_dma_stop()`. After that patch this atomicity is broken. Nested calls to `irq_local_disable()` or `irq_local_enable()` are fine as long as they're done with *different `flags`*. Those names are a bit misleading, in the kernel they're called `local_irq_save()` and `local_irq_restore()`. But the important point is, that if you call `irq_local_disable()` twice with different flags, if interrupts are enabled during the first call, it will save the CPU interrupt state with enabled interrupts into flags and disable interrupts. The next call to `irq_local_disable()` will save already disabled interrupts into a different copy of flags. Then the first call to `irq_local_enable()` with the second copy of the flags will restore that CPU state, which still had interrupts disabled. And only the second call to `irq_local_enable()` with the first copy of flags will re-enable the interrupts.